interfaces/builtin/cups.go

```diff
@@ -48,15 +58,108 @@ const cupsConnectedPlugAppArmor = `
+	// providing the cups socket
+
+	var cupsdSocketSource string
+	if err := slot.Attr("cups-socket", &cupsdSocketSource); err != nil {
+		return err
+	}
+
+	// TODO: we probably want to allow the empty slot for old cups snaps which
+	//       won't have the slot and have a sensible default for those, or
+	//       possibly just don't emit the mount if the attribute is not present
+
+	if cupsdSocketSource == "" {
+		return fmt.Errorf("cups slot must specify the location of the cups socket")
+	}
+
+	// TODO: should we do some checks on the cups-socket file location that it
+	//       is a clean filepath, etc.
+
+	return nil
+}
+
+func (iface *cupsInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	// add the base snippet
+	spec.AddSnippet(cupsConnectedPlugAppArmor)
+
+	// setup the snap-update-ns rules for bind mounting
+	emit := spec.AddUpdateNSf
+
+	var cupsdSocketSource string
+	err := slot.Attr("cups-socket", &cupsdSocketSource)
+	if err != nil {
+		return fmt.Errorf("broken interface: %v", err)
+	}
+
+	cupsdSocketSource = resolveSpecialVariable(cupsdSocketSource, slot.Snap())
+
+	emit("  # Mount cupsd socket from cups snap to client snap\n")
+	emit("  mount options=(bind rw) %s -> /var/cups/cups.sock,\n", cupsdSocketSource)
+	emit("  umount /var/cups/cups.sock,\n")
+
+	apparmor.GenWritableProfile(emit, cupsdSocketSource, 1)
+	apparmor.GenWritableProfile(emit, "/var/cups/cups.sock", 1)
+
+	// TODO: figure out why this bit is necessary - without this, apparmor
+	// denies trying to connect to the unix socket at /var/run/cups.sock with:
+	//
+	// apparmor="DENIED" operation="connect"
+	// profile="snap.test-snapd-cups-consumer.bin"
+	// name="/var/snap/test-snapd-cups-provider/common/cups.sock"
+	// pid=3195747 comm="nc" requested_mask="wr" denied_mask="wr" fsuid=0 ouid=0
+
+	spec.AddSnippet(fmt.Sprintf("%s rw,", cupsdSocketSource))
+
+	return nil
+}
+
+func (iface *cupsInterface) MountConnectedPlug(spec *mount.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	// get the source directory for the bind mount
+	var cupsdSocketSource string
+	err := slot.Attr("cups-socket", &cupsdSocketSource)
+	if err != nil {
+		// should be impossible since we should fail this in the sanitize stage
+		return fmt.Errorf("broken interface: %v", err)
+	}
+
+	cupsdSocketSource = resolveSpecialVariable(cupsdSocketSource, slot.Snap())
+
+	return spec.AddMountEntry(osutil.MountEntry{
+		Name: cupsdSocketSource,
+		Dir:  "/var/cups/cups.sock",
+		// the cups.sock file we are mounting over is very likely to exist
+		// already and it is also very likely to be a socket file, so we need to
+		// inform snap-update-ns that it is okay to perform our mount on top of
+		// the existing file only even if the existing file is a socket file
+		Options: []string{"bind", "rw", osutil.XSnapdKindFile(), osutil.XSnapdAllowSocketFile()},
+	})
```

Comment:
And one other thing: I think it would be preferable to have a read-only bind mount if at all possible. The `x11` interface does that for its mount of the `/tmp/.X11-unix` socket directory, so I'd assume this could work for the CUPS socket directory too.

Reply:
I thought about trying that but was hoping to get some feedback from the security team about why this denial is happening, but since I still haven't heard back about that I will give this a try instead

Reply:
Thinking more about this, I think we really do want to mount a directory rather than the socket directly even if you hadn't run into the these other issues.  Consider the following configuration:

1. one snap providing a `cups` slot with `cups-socket: $SNAP_COMMON/cups/cups.sock`, and defining a system daemon that will create the socket when it starts up.
2. a second snap with a `cups` plug that also provides a system daemon (e.g. this could be a Samba snap exposing the printers via SMB).

With the interface connected between these two snaps, reboot the system: the two system daemons will start in some indeterminate order (or possibly simultaneously). If the slot-side daemon starts first, then the socket will be created and be ready when we try to create the mount namespace for the plug-side daemon.

If the plug-side daemon starts first, it will notice that the socket file is missing and create an empty file in its place to use as a mount source (via the `ensureSource` method).  When the slot-side daemon starts, it will delete this file to make way for the socket it wants to bind. Now the slot-side daemon has a socket and the plug-side daemon has an empty file.

If we instead were to bind mount the directory containing the socket, start up order doesn't matter. If the plug-side daemon starts first, it will create and mount the directory where the socket will eventually exist. So when the slot-side daemon starts, its socket will be visible to the other snap.

Reply:
I think we should at least try/explore @jhenstridge suggestion here

Reply:
I changed the interface to provide a directory via the attribute and mount the directory instead.